k2/csrc/weights.cc

```diff
@@ -68,11 +69,83 @@ WfsaWithFbWeights::WfsaWithFbWeights(const Fsa &fsa, const float *arc_weights,
     : fsa(fsa), arc_weights(arc_weights), weight_type(t) {
   if (IsEmpty(fsa)) return;
   CheckInput(fsa, arc_weights);
+  ComputeForwardWeights();
+  ComputeBackardWeights();
+}
 
+// Mohri, M. 2002. Semiring framework and algorithms for shortest-distance
+// problems, Journal of Automata, Languages and Combinatorics 7(3): 321-350,
+// 2002.
+void WfsaWithFbWeights::ComputeForwardWeights() {
   auto num_states = fsa.NumStates();
-  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  backward_state_weights = std::unique_ptr<double[]>(new double[num_states]());
-  // TODO(haowen): compute `forward/backward_state_weights`, also check
-  // `IsTopSorted(fsa)`
+  forward_state_weights = std::unique_ptr<double[]>(new double[num_states]);
+  std::fill_n(forward_state_weights.get(), num_states, kDoubleNegativeInfinity);
+  std::vector<double> rd(num_states, kDoubleNegativeInfinity);
+
+  const auto &arcs = fsa.arcs;
+  forward_state_weights[0] = 0;
+  rd[0] = 0;
+  for (std::size_t i = 0; i != arcs.size(); ++i) {
+    const auto &arc = arcs[i];
+    DCHECK_GE(arc.dest_state, arc.src_state);
+    auto src_weight = forward_state_weights[arc.src_state];
+    auto &dest_weight = forward_state_weights[arc.dest_state];
+
+    // will be `kNegativeInfinity` for non-accessible state
+    double r = rd[arc.src_state] + arc_weights[i];
+    // will be `dest_weight` for non-accessible state
+    double d = AddWeights(dest_weight, r);
+    if (!ApproxEqual(dest_weight, d, kShortestDelta)) {
+      dest_weight = d;
```

Comment:
? I don't see an iteration here.
For now we only need this for the acyclic, top-sorted case; there is no need for the delta stuff in that case.

Reply:
Oh, ok, will delete